src/metabase/models/dependency.clj

```diff
@@ -0,0 +1,66 @@
+(ns metabase.models.dependency
+  (:require [clojure.set :as set]
+            [korma.core :as k]
+            [metabase.db :as db]
+            (metabase.models [interface :refer :all])
+            [metabase.util :as u]))
+
+;;; # IRevisioned Protocl
+
+(defprotocol IDependent
+  "Methods an entity may optionally implement to control how dependencies of an instance are captured."
+  (dependencies [this id instance]
+    "Provide a map of dependent models and their corresponding IDs for the given instance.  Each key in the returned map
+     must correspond to a valid Metabase entity model otherwise it will be ignored.  Each value for a given key should
+     be a collection of unique integer ids for the given model.
+
+     For example:
+         (dependencies Card 13 {})  ->  {:segment [25 134 344]
+                                         :table   [18]}"))
+
+
+;;; # Dependency Entity
+
+(defentity Dependency
+  [(k/table :dependency)])
+
+
+;;; ## Persistence Functions
+
+
+(defn retrieve-dependencies
+  "Get the list of dependencies for a given object."
+  [entity id]
+  {:pre [(metabase-entity? entity)
+         (integer? id)]}
+  (db/sel :many Dependency :model (:name entity) :model_id id))
+
+(defn update-dependencies
+  "Update the set of `Dependency` objects for a given entity."
+  [entity id deps]
+  {:pre [(metabase-entity? entity)
+         (integer? id)
+         (map? deps)]}
+  (let [entity-name      (:name entity)
+        dependency-set   (fn [k]
+                           ;; TODO: validate that key is a valid entity model
+                           (when (every? integer? (k deps))
+                             (for [val (k deps)]
+                               {:dependent_on_model (name k), :dependent_on_id val})))
+        dependencies-old (set (db/sel :many :fields [Dependency :dependent_on_model :dependent_on_id] :model entity-name :model_id id))
+        dependencies-new (->> (mapv dependency-set (keys deps))
+                              (filter identity)
+                              flatten
+                              set)
```

Comment:
Could use `for` loop here. Not sure if this is exactly your logic but something like this:
```clojure
(set (for [k       (keys deps)
           dep-set (dependency-set k)
           :when   dep-set
           v       dep-set]
       v))
```

`for` loops can do all sorts of magical things.
```clojure
(let [k->v {:a [1 2 3]
            :b [4 5 6]
            :c [7 8 9]}]
  (for [k [:a :c]
        v (k->v k)
        :when (not (zero? (mod v 3)))]
    v))
;; -> (1 2 7 8)
```